### DISCUSSION TITLE
Classic resets - strategy

### DISCUSSION THREAD

**Discussion by RandomEC** [evennia/evennia#2635] [General Discussion]
So, I'm planning for my MUD to have sort of the traditional Diku/Merc/Envy style resets. Every so often, the MUD checks and puts each area back to its starting configuration.

My basic plan is to send mobiles to location "None" when the die (so apart from creating a mob for the first time, there should never be a need to load again), and objects that get sacrificed or eventually fall to dust on the floor will be treated the same. Since objects can be obtained and kept by the players, however, there will need to be some mechanism for them to be reloaded. I've been thinking about this a bunch, and have come up with two general plans for how to handle this, and am having difficulty telling which would be preferable. Any input is appreciated.

**Method 1: Handle the whole thing with scripts.** In this plan, there would be two types of scripts - 1) an overarching one that has all the information on how to do each type of reset, and would handle when each area gets reset, and 2) each area would have its own script that would hold the data on how that area gets reset. The primary downside that I see to this plan is that it feels like this would be more difficult to do organically, as you build, than the below method - though it would work quite nicely for pre-existing areas.

**Method 2: Store reset information on rooms/mobs/objects.** In this plan, global resetting would still be handled by an overall script,  which would call an "at_reset" hook on everything tagged with the area name. Each room/mobile/object/door would carry the relevant information for some resets, and would have a method that gets called at_reset for effectuating the reset as part of its class, as follows:
     Mobiles - would carry information about resetting themselves, and objects that they either wear or have in inventory.
     Rooms - would carry information about objects that reset in that room, not on mobiles or in containers.
     Objects - containers would carry information about objects that reset in them.
     Doors - would carry information on what their reset state is.
This feels like it would be easier to do organically as you build. The only real downside that I can see is that this could cause a slowdown when everything tries to reset at the same time, as opposed to the first method, which would be sequential, and go reset by reset.

It feels like the work on implementing either system is probably about the same to me, is there any computational or coding pitfall that I might be overlooking that would make one substantially preferable over the other?

Thanks!

**Comment by Tegiminis** [comment]
Under either method, you should definitely store all information and the reset method on the object (for example: `home`, `maxhealth`, etc)  for the reason you laid out (easier for builder/developers to fiddle with, without creating custom scripts for every reset). As for how that reset method gets called, zone scripts does seem like the most robust way. You could also circumvent scripts and have objects subscribe to the ticker handler. 

I'm doing something a little silly myself, where mobs that die go to a room locked to players instead of "None", so that builders have a little workshop for messing with NPCs in each zone. This also means that resetting NPCs is very simple, as the script only has to be aware of the room the NPCs are in to function, rather than the whole zone.

I wouldn't worry about being computationally intensive at this point in time. Architect the system so that it has the least amount of friction for you, the developer, and ideally builders. You can always refactor your code later if it turns out to be onerous. The key here is to make sure you divide your code into data (all information relevant to the reset), function (the actual reset call, what it does with that data), and event (the "trigger" for your reset, calls the function with the data), so that refactoring it is less of a pain should you need to do so.

**Reply by RandomEC** [reply]
> Under either method, you should definitely store all information and the reset method on the object (for example: `home`, `maxhealth`, etc) for the reason you laid out (easier for builder/developers to fiddle with, without creating custom scripts for every reset). As for how that reset method gets called, zone scripts does seem like the most robust way. You could also circumvent scripts and have objects subscribe to the ticker handler.

Agreed. Certainly the data for the mobiles resetting will be kept on them, since there is no (well, barring me doing something stupid) risk that they will need to load again after the first time. For objects, though, which potentially will have to be loaded again (e.g. player beats skeleton, skeleton goes to none, player takes the sword from his corpse and keeps and uses it. When skeleton resets, he has no sword and one needs to load), my question is where to keep the data on that object so it can be loaded. I think the obvious two choices are 1) on the object (mobile, room or container object) where it should be reset, or 2) in some central script where all reset information of that sort is kept. Since, as you point out, at least SOME of the relevant reset information ("home") is on the object necessarily, it seems like this is probably the way to go. 

I did think instead of having each object subscribe to the ticker handler, but keeping track of one overall zone ticker rather than (potentially) hundreds per zone just sounded better. 

> 
> I'm doing something a little silly myself, where mobs that die go to a room locked to players instead of "None", so that builders have a little workshop for messing with NPCs in each zone. This also means that resetting NPCs is very simple, as the script only has to be aware of the room the NPCs are in to function, rather than the whole zone.
> 
> I wouldn't worry about being computationally intensive at this point in time. Architect the system so that it has the least amount of friction for you, the developer, and ideally builders. You can always refactor your code later if it turns out to be onerous. The key here is to make sure you divide your code into data (all information relevant to the reset), function (the actual reset call, what it does with that data), and event (the "trigger" for your reset, calls the function with the data), so that refactoring it is less of a pain should you need to do so.

Good advice, thanks!

I think that where I'm going to come down on this, based on our discussion is as follows, using the above skeleton example:

Skeleton will, obviously, have his "home" information (he'll also have max hp info, but that will be reset at the end of combat before he goes to "none"). He will also have all the relevant information for sword in a dictionary of dictionaries. Finally, he will have an "at_reset" method on him that implements moving him to home, and either a) retrieving the object from "none" and equipping, or b) loading and equipping the object, as, apart from loading the object, those are all things that happen to/on him. So, in your final paragraph, the data and function will be on each object.

Then, centrally, I will have a script that is always running that checks for whether an area needs to be reset or not at a certain interval (areas with no players in them reset more often), and, if so, will call the "at_reset" hook on all objects with that zone tag. This will be the event. 

Thanks again for all your help!

**Reply by Tegiminis** [reply]
You're welcome! I enjoy talking about this sort of architecting :)

For the skeleton sword problem, I actually have a different pitch for you.

First, you absolutely can store the item reference on the skeleton, and create a new one on reset. But this introduces some extra steps to consider. For example, we must check if the skeleton has the sword already when the reset happens, and have the AI run the equip command after reset. Also, if you decide to change some stat on the weapon, you will have to reinstantiate all weapons of that type, like through a global script or a custom builder command.

You might consider instead separating "player weapons" from "NPC weapons". Instead of having the skeleton use an object like a player for combat, it instead references a class that is never instantiated into worldspace. This has a number of distinct benefits:

1. You can change the weapon's balance and have it immediately reflected across all instances without having to reinstantiate them; just requires a server reload.
2. You have more control over what the mob drops. Maybe you want to give this skeleton a really powerful sword, but you don't want the players to have it, for example.
3. You keep all your NPC information self-contained, so the reset process is simpler and requires less code

The main drawback is that unless you engineer your items and rules in a very careful, threading-the-needle way, you will have to write separate functions for player and NPC attacks. This increases code complexity, which means more maintenance should you decide to change the rules. Making AI and players follow the exact same rules to the letter is also aesthetically pleasing and immersive.

Neither method is "correct", but all down to what you want out of your game, and what you feel will be easiest for you as a builder.

**Reply by RandomEC** [reply]
As a practical matter, I'm employing the old-school MUD cheat of SUPER abstracting monster damage, based solely on level, with a multiplier of 1.5 if they are wielding a weapon. So the actual nature of the weapon doesn't matter a huge amount. My primary plan is to get something up and running as an absolute first priority, so it is probably going to look pretty basic and Envy-y. You probably won't even have to squint very hard to see it. Then I'm going to decide what direction I want to go to make more fancy from there.

I had the same thoughts as you as far as checking if he already has the weapon and equipping it, but that doesn't feel like too heavy of a lift, so I think I'm probably just going to stick with that. Appreciate the suggestion, though, and I love talking about this stuff, too. :)

**Reply by RandomEC** [reply]
For anyone to whom this is relevant, here's what I ended up doing:

Pre-resetting analysis:

1. I used the suggestion of some others that mobiles not be destroyed when they "die", rather just have their location set to "none".
2. Objects are a bit more fiddly. Objects that either are sacrificed, or disintegrate from laying on the ground too long can be treated like mobiles, above. But, obviously, the overall number of an object on the mud will need to be able to increase as players pick up and keep objects.
3. Doors/exits are the only other thing that reset, and they aren't going anywhere.

Reset implementation
1. I decided to place an at_reset method on both mobiles and exits, and on rooms, for a reason that will be explained in a moment. 
2. All the mobiles, exits and rooms in an area are then tagged with a tag of the area name in the category "area names".
3. A global script is the central repository for all the area names, and at a predetermined amount of time checks to see if each area needs to be reset (5 minutes have passed for areas unoccupied by players, 15 for areas occupied by players) and, if so, finds everything with that area name tag, and iterates through calling the at_reset method on each of them. The script repeats every five minutes, and has a method to check whether there is a player in the area, and a counter for each area which it increments if there is a player in the area. 
4. The mobile at_reset method simply (quietly) returns the mobile to its home room, and checks the mobile to see if all the objects that should be on it are. If not, it checks "none" for those objects (and brings them to the mobile if so), or, failing that, finds a copy of the item on the mud and makes a copy of it, and brings that to the mobile.
5. The room at_reset method simply checks whether objects that reset in the room (not on mobiles) are present, and follows the same approach as mobiles, above, for getting them there. I put this on the room rather than the object, as it made more sense to have ONE room trying to reset the object, rather than 10 copies of an object figuring out which of them was going to do the resetting. The room method also sends a message to everyone in the room so that they know that the area has reset.
6. The exit at_reset method simply copies its default door state over whatever the current door state is.

Hope that's useful for someone!

**Comment by telimektar3** [comment]
When the mobile resets and checks for its objects, is it potentially going to set objects picked up by players to none as well? Sorry if I’m not following.

**Reply by RandomEC** [reply]
Sorry, Tim, probably I wasn't clear enough. Here's a little more detail on how a mobile handles what objects it should have on reset.

1. It checks to see if it already has equipped and/or in its inventory the objects that should be there. If so, it goes no further.
2. If not, it first checks to see if any of the objects that it should have currently exist with a location of "none". Any objects currently possessed by players should have a location of that player. The only way they should have a location of "none" is if no one is using them. The mobile itself never sets any object's location to "none", it just looks for items that have a location of "none" already. If it finds objects at none, the objects' location is changed to the mobile, who then equips them, if necessary.
3. If the mobile STILL doesn't have the objects it needs, it looks for the items elsewhere on the MUD. Once it finds one, it makes a copy of it, moves the copy to the mob, then clears any changes that may have been made to the item (enchantments, etc. are set in a separate dictionary to enable this, keeping the core identity of the object intact). Then it equips it if necessary.

**Reply by telimektar3** [reply]
Yes, this added detail makes sense to me.
